### PR TITLE
Use of static instead of ClassName

### PR DESCRIPTION
**Pull request by MathiasReker**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Inside class or interface element self should be preferred to the class name itself.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | Does it break backward compatibility? no
| Deprecations? | Does it deprecate an existing feature? no
| Fixed ticket? | /
| How to test?  | /

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11730)

### CODE REVIEW

**Review by PierreRambaud — DISMISSED** [review]
@PrestaShop/prestashop-core-developers No need QA.

File: var/SymfonyRequirements.php
Comment (jolelievre) [inline var/SymfonyRequirements.php]: I wouldn't touch to this file as it will change with Symfony upgrade

File: src/Core/Localization/DataLayer/CurrencyDataLayerInterface.php
Comment (jolelievre) [inline src/Core/Localization/DataLayer/CurrencyDataLayerInterface.php]: As this is an interface I think it's better to keep the explicit Interface in the prototype

File: src/Core/Localization/DataLayer/LocaleDataLayerInterface.php
Comment (jolelievre) [inline src/Core/Localization/DataLayer/LocaleDataLayerInterface.php]: As this is an interface I think it's better to keep the explicit Interface in the prototype

**Review by jolelievre — DISMISSED** [review]

**Review by jolelievre — CHANGES_REQUESTED** [review]
oups I clicked the wrong options, this needs a few changes

**Comment by MathiasReker** [comment]
Thanks

**Review by jolelievre — DISMISSED** [review]

**Comment by jolelievre** [comment]
@PierreRambaud is it ok for you?

File: src/Core/Localization/CLDR/LocaleData.php
Comment (PierreRambaud) [inline src/Core/Localization/CLDR/LocaleData.php]: Keep LocaleData

File: src/Core/Localization/CLDR/NumberSymbolsData.php
Comment (PierreRambaud) [inline src/Core/Localization/CLDR/NumberSymbolsData.php]: keep NumberSymfolsData

File: src/Core/Localization/Currency/CurrencyCollection.php
Comment (PierreRambaud) [inline src/Core/Localization/Currency/CurrencyCollection.php]: Keep CurrencyCollection

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Need to change methods signatures

**Review by PierreRambaud — APPROVED** [review]

**Comment by jolelievre** [comment]
@PierreRambaud just out of curiosity, when do you consider `self` acceptable in some signatures rather than others?

**Comment by MathiasReker** [comment]
Ready to get merged I think :)

File: src/Core/Cart/AmountImmutable.php
Comment (PierreRambaud) [inline src/Core/Cart/AmountImmutable.php]: Need change

File: src/Core/Cart/AmountImmutable.php
Comment (PierreRambaud) [inline src/Core/Cart/AmountImmutable.php]: Need change

File: src/Core/Domain/Category/ValueObject/CategoryId.php
Comment (PierreRambaud) [inline src/Core/Domain/Category/ValueObject/CategoryId.php]: Need change

File: src/Core/Domain/Category/ValueObject/CategoryStatus.php
Comment (PierreRambaud) [inline src/Core/Domain/Category/ValueObject/CategoryStatus.php]: Need change

File: tools/build/Library/InstallUnpacker/classes/VersionNumber.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: Need change

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Need change

**Comment by MathiasReker** [comment]
@PierreRambaud solved

**Comment by MathiasReker** [comment]
Ready to get merged? :)

**Review by jolelievre — DISMISSED** [review]
Seems good to me, another approval from @PrestaShop/prestashop-core-developers before merging please

**Review by eternoendless — CHANGES_REQUESTED** [review]
This will break the overrides system.

All classes inside the `classes` directory should not use `self`.

Non overridable classes should use `static` instead of `self`.

**Comment by MathiasReker** [comment]
I'm not sure which classes are not overridable?

**Comment by eternoendless** [comment]
Just use `static` everywhere and you'll be safe.

**Comment by MathiasReker** [comment]
I think it is OK now

**Review by jolelievre — DISMISSED** [review]
@MathiasReker you need to use static everywhere instead of self, this is required to allow classes override with late static binding, more info here: https://www.leaseweb.com/labs/2014/04/static-versus-self-php/

**Comment by jolelievre** [comment]
You can also rename the PR "Use of static instead of ClassName"

**Comment by MathiasReker** [comment]
@jolelievre thanks for clarifying. I changed most of self to static. I let two places be self, because self was already used in same class.

**Comment by jolelievre** [comment]
> @jolelievre thanks for clarifying. I changed most of self to static. I let two places be self, because self was already used in same class.

It was probably an error to use self in the first place, @eternoendless wdyt? should we replace all `self` with `static` ?

**Comment by MathiasReker** [comment]
@jolelievre would you like to use static instead of self for for the whole project?

**Comment by jolelievre** [comment]
I'm not sure, I just discovered the difference between the two, so I'd prefer to have @eternoendless ' opinion

**Comment by Quetzacoalt91** [comment]
Be careful, I wonder how it works when an override exists for these class.

**Comment by MathiasReker** [comment]
Conflicts solved.

**Comment by eternoendless** [comment]
> Be careful, I wonder how it works when an override exists for these class.

An override is a child class that extends a core class: `Product extends ProductCore`.

Unless there's a mistake in the code, `ProductCore` should never be referenced directly, except when extending it. Instead, we always use it as `Product`.

The [late static binding feature](http://php.net/manual/en/language.oop5.late-static-bindings.php) means that if we call `Product::doSomething()` statically, and `doSomething()` calls `static::doOtherThing()`, then PHP will call the implementation of `doOtherThing()` from the child class instead of the parent class.

This means that using `static` should be safe for overrides as long as we never call `ProductCore` directly.

**Comment by eternoendless** [comment]
I rebased and squashed all commits from this branch. 

@MathiasReker in the future please rebase instead of merging develop into your branch. This keeps the branch history cleaner. If you don't know how to do it, [here's how](http://build.prestashop.com/howtos/misc/rebasing-your-git-branch/).

The squashing is optional, it depends if there are many things done in your PR. If it's mostly about changes required by the team, then it's better to squash those changes into a single commit.

**Review by eternoendless — DISMISSED** [review]

**Comment by MathiasReker** [comment]
all good now?

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
LGTM wait for someone else to review

**Review by eternoendless — COMMENTED** [review]
LGTM

**Comment by PierreRambaud** [comment]
Thanks @MathiasReker

**Pull request merged**
